tachyon/zk/plonk/prover/synthesizer.h

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_PLONK_PROVER_SYNTHESIZER_H_
+#define TACHYON_ZK_PLONK_PROVER_SYNTHESIZER_H_
+
+#include <utility>
+#include <vector>
+
+#include "absl/numeric/bits.h"
+
+#include "tachyon/zk/base/entities/prover.h"
+#include "tachyon/zk/plonk/constraint_system.h"
+#include "tachyon/zk/plonk/prover/witness_collection.h"
+
+namespace tachyon::zk {
+
+template <typename PCSTy>
+class Synthesizer {
+ public:
+  constexpr static size_t kMaxDegree = PCSTy::kMaxDegree;
+  constexpr static size_t kDomainSize = kMaxDegree + 1;
+
+  using F = typename PCSTy::Field;
+  using Poly = typename PCSTy::Poly;
+  using Evals = typename PCSTy::Evals;
+  using RationalEvals = typename PCSTy::RationalEvals;
+
+  Synthesizer() = default;
+  Synthesizer(size_t num_circuits, const ConstraintSystem<F>* constraint_system)
+      : num_circuits_(num_circuits), constraint_system_(constraint_system) {
+    challenges_.reserve(constraint_system->num_challenges());
+
+    advice_values_.reserve(num_circuits_);
+    advice_blinds_.reserve(num_circuits_);
+    for (size_t i = 0; i < num_circuits_; ++i) {
+      // And these may be assigned with random order.
+      advice_values_.resize(constraint_system->num_advice_columns());
+      advice_blinds_.resize(constraint_system->num_advice_columns());
+    }
+  }
+
+  // Synthesize circuit and store advice columns.
+  template <typename CircuitTy>
+  void GenerateAdviceValues(
+      Prover<PCSTy>* prover, const std::vector<CircuitTy>& circuits,
+      const std::vector<std::vector<Evals>>& instance_value_table) {
```

Comment:
1. A User calls `CrateProof()` with selected `instance_columns_vec`.
2. `Synthesizer` receives `instance_columns_vec` and build `advice_columns_vec`.
3. `Argument` receives `instance_columns_vec` and `advice_columns_vec` and store them.
4. When lookup and permutation request some type of `Table`, `Argument` exports it to the caller.
This code is in second step. Therefore the type of argument is correct.
However the name of argument is weird. It will be fixed!